db/view/view.cc

```diff
@@ -1770,12 +1770,18 @@ future<> view_update_generator::mutate_MV(
             ++stats.view_updates_pushed_local;
             ++cf_stats.total_view_updates_pushed_local;
             ++stats.writes;
+            size_t units_cnt = sem_units.count();
+            if (!remote_endpoints.empty()) {
+                // We still have pending remote endpoints to send to, so the semaphore units should be taken by the longer,
+                // remote update and not the local one
+                units_cnt = 0;
+            }
```

Comment:
Given the small size of the PR, here is my obligatory nitpick: how about using the ternary operator:

```c++
const size_t units_cnt = remote_endpoints.empty() ? sem_units.count() : 0;
```

This way we can make the variable const.

Reply:
As for the correctness: I know that remote update _usually_ should take longer than the local update, but nothing prevents an opposite situation from happening. Shouldn't we hold on to the units until both remote and local update finishes?

That's at least what I _think_ should happen, given the description in the issue and the PR; I'm not familiar with the code enough yet to tell whether the whole approach makes sense or not.

Reply:
Yes, I have the same question - why make any assumption what takes longer? Why not just hold the units until both happen?

Note that we have two code paths - one does synchronous view updates:
```
co_await when_all_succeed(
                    std::move(local_view_update), std::move(remote_view_update))
.discard_result();
```
the other only waits for the local view update:
```
(void)remote_view_update;
co_return co_await std::move(local_view_update);
```
In the second case I think you'll need to modify the code to do something like (completely untested)
```
co_await std::move(local_view_update);
// don't wait for remote_view_update, but only release the semaphore units after it (and local_view_update) finished
(void) remove_view_update.finally([std::move(units)]);
co_return;
```

Reply:
If we're nitpicking, the variable should be named to reflect its role, not its type/contents.

This isn't just a units count, it has a special role given the split below. If the name could increase the understanding of what it does, it would be a benefit.

For example, semaphore_units_to_hold_for_local_view_update.

We could even have two variables of type semaphore_units, and move the split here, instead of splitting (ugh) the logic between the decision point here and the split point a few lines below. The split point would read units = std::move(local_update_semaphore_units).

Reply:
btw it would be preferable not to split 0/all, so if the local update happens to take more, it imposes something on the semaphore.